sky/jobs/scheduler.py

```diff
@@ -66,129 +66,134 @@
-                    dag_yaml_path = maybe_next_job['dag_yaml_path']
-                    env_file_path = maybe_next_job['env_file_path']
-
-                    _start_controller(job_id, dag_yaml_path, env_file_path)
 
+    # this should have been made by now
+    if not os.path.exists(JOB_CONTROLLER_ENV_PATH):
+        return
+
+    logs_dir = os.path.expanduser(
+        managed_job_constants.JOBS_CONTROLLER_LOGS_DIR)
+    os.makedirs(logs_dir, exist_ok=True)
+    log_path = os.path.join(logs_dir, 'controller.log')
+
+    with open(JOB_CONTROLLER_ENV_PATH, 'r', encoding='utf-8') as f:
+        run_cmd = f.read()
+
+    pid = subprocess_utils.launch_new_process_tree(run_cmd, log_output=log_path)
+    with open(JOB_CONTROLLER_PID_PATH, 'a', encoding='utf-8') as f:
+        f.write(str(pid) + '\n')
+
+
+def get_alive_controllers() -> typing.Optional[int]:
+    if not os.path.exists(JOB_CONTROLLER_PID_PATH):
+        # if the file doesn't exist, it means the controller server is not
+        # running, so we return 0
+        return 0
+
+    try:
+        with open(JOB_CONTROLLER_PID_PATH, 'r', encoding='utf-8') as f:
+            pids = f.read().split('\n')[:-1]
+    except OSError:
+        # if the file is corrupted, or any issues with reading it, we just
+        # return None to be safe and not over start
+        return None
+
+    alive = 0
+    for pid in pids:
+        try:
+            if subprocess_utils.is_process_alive(int(pid.strip())):
+                alive += 1
+        except ValueError:
+            # if the pid is not an integer, let's assume it's alive to not
+            # over start new processes
+            alive += 1
+    return alive
+
+
+def maybe_start_controllers(env_file_path: typing.Optional[str] = None) -> None:
+    """Start the job controller process.
+
+    If the process is already running, it will not start a new one.
+    Will also add the job_id, dag_yaml_path, and env_file_path to the
+    controllers list of processes.
+    """
+    try:
+        with filelock.FileLock(JOB_CONTROLLER_PID_LOCK, blocking=False):
+            alive = get_alive_controllers()
+            if alive is None:
+                return
+            wanted = get_number_of_controllers()
+            started = 0
+
+            if wanted != alive and env_file_path:
+                activate_python_env_cmd = (
+                    f'{constants.ACTIVATE_SKY_REMOTE_PYTHON_ENV};')
+                source_environment_cmd = (f'source {env_file_path};'
+                                          if env_file_path else '')
+                run_controller_cmd = (f'{sys.executable} -u -m'
+                                      'sky.jobs.controller')
+
+                run_cmd = (f'{activate_python_env_cmd}'
+                           f'{source_environment_cmd}'
+                           f'{run_controller_cmd}')
+
+                with open(JOB_CONTROLLER_ENV_PATH, 'w', encoding='utf-8') as f:
+                    f.write(run_cmd)
```

Comment:
Wait, why do we write this to a file? These are all constants except env_file_path - we can just pass that down into `start_controller` and build the string there, like before.